### PR TITLE
perf(pgen): cut the single-partition scan from 11.2s to 1.19s

### PR DESCRIPTION
**Pull request by mwiewior**
## Summary

A whole-chromosome PGEN scan spent most of its time turning decoded genotypes into Arrow arrays rather than decoding them. This branch removes that overhead layer by layer, ending with a fused decode for the record type that dominates a `plink2 --make-pgen` fileset.

Benchmark throughout: chromosome 22 of 1000 Genomes, 993,881 variants × 2,548 samples = **2,532,408,788 genotypes**, single partition, release build with `-C target-cpu=native`.

The Rust-only scan, no Python and no materialization (`examples/pgen_ds_profile`):

| field | at branch point | now | |
|---|---:|---:|---:|
| `DS` (float32) | ~11.2 s | **1.19 s** | 9.4× |
| `ALT_COUNT` (int8) | — | **0.59 s** | new column |

## What changed

1. **Arrow values/validity buffers built directly** instead of a per-cell `ListBuilder::append_option`. Isolated bench: 352 → 860 Melem/s.
2. **`DS` joined the single-field fast path `GT` already had**, removing a per-variant `Vec<Option<f32>>` — eight bytes per cell and one allocation per variant.
3. **`append_codes` rewritten table-driven** with bulk validity.
4. **Hardcall phase orientation skipped for dosage projections.** A phased heterozygote and an unphased one carry the same dosage, so applying it was a full pass over every sample of every phased record for no change in output.
5. **`ALT_COUNT` column added** — the hardcall ALT allele count as `Int8`, one byte per genotype instead of the four `DS` needs. 2.53 GB of output instead of 10.13 GB on this chromosome.
6. **Expansion loops made auto-vectorizable.** The ALT count for codes 0..=4 is `0,1,2,0,1`, which is `code - 3 * (code >= 3)` — no lookup table, no `Vec::push`, so LLVM emits NEON.
7. **Difflist buffer reused** across variants instead of allocated per record (~10⁶ allocations per chromosome).
8. **The common-value + difflist branch fused into the output buffer** — the largest single win, below.
9. **Output shape reported in the schema metadata**, so a caller can preallocate.
10. **The `.pvar` parsed across threads** — the scan's serial prologue, below.

## The fused decode (8)

81% of a `plink2 --make-pgen` fileset is main-track representation 4, 6 or 7: one common category for every sample plus a sparse difflist of exceptions. Unlike every other representation it has **no per-sample base to unpack**, so the two passes it went through — fill a `Vec<u8>` of codes, then read that buffer back to write the `f32`/`i8` — were one pass more than the record needs.

`DS` and `ALT_COUNT` now fill their Arrow values slice from the common category and patch the difflist straight into it. Validity is one run for the whole variant plus a bit per patch, so bitmap work is proportional to the difflist rather than the sample count.

| field | before | after | |
|---|---:|---:|---:|
| `DS` | 2.31 s | **1.19 s** | 1.94× |
| `ALT_COUNT` | 1.65 s | **0.59 s** | 2.80× |

**The obstacle was the hardcall-phase track.** It must still be validated and consumed — its length depends on the heterozygote count, so misparsing it would silently shift every following track — but that count came from scanning the per-sample buffer fusion removes. It is derivable from the difflist alone: every unpatched sample carries the common category, so it is heterozygous exactly when that category is. Phase validation is now shared with the dense path via `validate_phase_track`.

Three exclusions, each load-bearing: `GT` (it distinguishes `(0,1)` from `(1,0)`, and the fused decode discards the orientation), LD bases (`retain_main` — those need the full main track anyway, ~13% of records), and anything with a dosage/HDS/multiallelic track.

### Issue #233 proposes the wrong optimization

#233 argues for keeping the decoded main track 2-bit packed the way pgenlib does. That reading came from the LD-compressed branch (13% of records). The dominant branch has no per-sample base to pack: pgenlib's `vecset` + `Expand2bitTo8` writes `sample_ct/4 + sample_ct` bytes, while a fused fill writes `sample_ct` and nothing else. Packing would make us **slower** here. Corrected in a comment on the issue.

## The parallel `.pvar` parse (10)

Opening a fileset parses the whole `.pvar` before any partition runs, so it is a
fixed serial prologue under every scan — on chromosome 22, 108 MB of text and
**0.257 s**, which was 20% of a four-partition read and did not shrink when
partitions were added. The body is now split on line boundaries and parsed on
scoped threads that borrow the text directly, so nothing is copied and no runtime
is involved. **0.257 s → 0.068 s, 3.8×.**

Two behaviours had to survive being parsed out of order:

- **Error messages carry file line numbers**, but a chunk only knows its own. Each
  chunk reports where it stopped in local lines and the caller adds the offset,
  counting newlines ahead of that chunk only on the error path.
- **The row limit is enforced before the offending line is read**, so on a file
  that is both over the limit and malformed, the limit wins — there is an
  existing test asserting exactly that. A chunk cannot see the running total, so
  it applies the limit to itself as a memory bound and the caller reconciles the
  counts as chunks are joined, taking whichever condition comes first in the
  file.

The chunk count is injectable so tests can pin the split; inheriting the host's
core count would make those boundary cases untestable. My first attempt at these
tests passed under mutation — deleting either half of the reconciliation left
them green, because they reached the same error by another route. The tests now
in the branch fail when either half is removed.

## Scaling

`examples/pgen_scaling_probe.rs` measures the scan against partition count with
no Python in the way, and reports `DependencyRecords` alongside — the count of
records a partition decodes to reconstruct an LD chain but never emits, which is
the obvious suspect for parallel work that duplicates rather than divides.

| Partitions | `DS` scan | speedup | `DependencyRecords` |
|---:|---:|---:|---:|
| 1 | 1.250 s | 1.00× | 0 |
| 4 | 0.386 s | 3.24× | 0 |
| 8 | 0.248 s | **5.05×** | 0 |

The scan divides cleanly and LD chains cost nothing. End-to-end scaling is much
lower (1.99× at four partitions) because the scan is only a third of a
materializing read; the rest is a copy that saturates memory bandwidth. That
analysis lives in `bioformats-benchmark/PGEN_BENCHMARK.md`.

## Correctness

- **`ALT_COUNT` and `DS` match the `GT` scan across all 2,532,408,788 cells of chr22.** `GT` never takes the fused path, so this is a real oracle on real records at full scale. New `examples/pgen_field_parity` does it; it exits nonzero on disagreement and was verified to fail on a deliberately corrupted patch.
- The **pgenlib and snputils differential oracles** stay green, run against the real libraries (`PGEN_REQUIRE_REFERENCE_ORACLES=1`), not the skip path.
- Unit tests compare the fused fill against the two-pass expansion **buffer for buffer and bitmap for bitmap**, and check the fused decode reconstructs exactly the main track the existing decoder produces.
- **Never narrow `DS` from `Float32`.** PGEN dosages are genuinely fractional; on a dosage fileset the two tracks disagree completely:
  ```
  dosages   : [ 0.125  1.0  1.875  missing ]
  hardcalls : [ missing  1  missing  missing ]
  ```
  An intermediate version of the fast path returned hardcalls where dosages were asked for, and the differential oracles caught it.

## End to end

Through polars-bio (which needs biodatageeks/polars-bio#436 for the matrix reader), all readers interleaved in one session, one thread each — pgenlib and snputils are serial:

| | polars-bio at branch point | now | snputils | pgenlib |
|---|---:|---:|---:|---:|
| dosage (float32) | 4.338 s | **1.285 s** | 3.558 s | 1.863 s |
| hardcall (int8) | 2.959 s | **0.689 s** | 1.538 s | 0.862 s |

**polars-bio is now faster than pgenlib on both workloads at equal core count** —
1.45× on dosage and 1.25× on hardcalls — having started this branch 2.46× and
3.56× slower. At four and eight partitions it is 3.65× and 4.75× pgenlib on dosage.

Two qualifications. pgenlib drifted by up to 4.6% between sessions earlier in this
branch, though not in the run these figures come from — it landed within 1% of
the previous one — so the margins stand as measured. Still prefer "faster" to a
third decimal place across sessions. And **pgenlib still uses slightly less memory** — 12.1 GB against 12.6 GB
on dosage — though that gap is down from 65% to 4%.

## Decoding into the caller's buffer

The scan builds Arrow batches and something else copies them into whatever the
caller actually wanted. For a dense matrix that copy is pure overhead — 10 GB
read and 10 GB written on a whole chromosome — and it is what stopped a
materializing read from scaling, because it saturates memory bandwidth long
before it runs out of threads.

`matrix::read_genotype_matrix` skips it: decoders write genotypes at their final
address, so the values are touched once, by the threads that produced them.
`GenotypeMatrixReader` holds the opened fileset across the shape query and the
decode, because a caller must learn the shape before it can allocate and asking
must not cost a second parse of the 108 MB PVAR — an earlier revision paid it
twice and briefly made the hardcall workload *slower* than the Arrow path.

Two things made this simpler here than in the Arrow path rather than harder. A
matrix has no validity bitmap, so a missing call is a sentinel the caller picks.
And a full-file read gives every variant a row index equal to its position in the
selection, so partitions own disjoint contiguous row ranges: the destination
splits with `split_at_mut` and workers never share a byte. Partition planning,
including LD base chains, is the existing planner's rather than a second
implementation.

The general decoder needed care the fast paths did not. It is the only branch
that sees a stored dosage track, and those values are genuinely fractional; an
earlier draft routed them through the internal-code mapping, which can only say
0, 1 or 2, so every dosage came out as zero while every fast-path fixture still
passed. There is a test on the fixture whose dosage and hardcall tracks
disagree, and it fails when that mapping is reinstated.

Tests compare the matrix against the scan cell for cell on the fixture holding
every record representation, at one, two, four and eight threads, plus sample
subsetting and a wrongly sized destination.

pgenlib and snputils reproduced their earlier figures to within 1% in the same session, so the deltas are the change and not machine drift. Zero bitwise differences against pgenlib across all 2,532,408,788 cells in both workloads, with the single-cell corruption self-test confirming the comparison can still fail.

### CODE REVIEW

**Comment by mwiewior** [comment]
@claude pls review

**Comment by mwiewior** [comment]
@codex pls review

File: datafusion/bio-format-pgen/src/physical_exec.rs [hunk 1]
```diff
@@ -596,7 +1272,7 @@ fn execute_gt_only(
             estimated_row_bytes,
         );
         let mut batch =
-            GtBatchBuilder::new(&schema, partition_row_capacity, selected_sample_count)?;
+            FastFieldBuilder::new(&field, &schema, partition_row_capacity, selected_sample_count)?;
```
[the remark this answers is not in the document]
Comment (mwiewior) [reply datafusion/bio-format-pgen/src/physical_exec.rs:1278]: Fixed in c6618ac.

`execute_single_field` now sizes from the field it was given (`std::slice::from_ref(&field)`), and `estimate_genotype_bytes` has an `ALT_COUNT` arm at two bytes per sample — one `i8` plus validity and the list offset — so a multi-field projection no longer prices it at zero either.

Measured on chromosome 22 (993,881 x 2,548): the `ALT_COUNT` scan builds **76 batches where it built 189**, at 0.60 s either way. A unit test pins the estimate, asserts it is below the `GT` one, and asserts the multi-field sum.

File: datafusion/bio-format-pgen/src/matrix.rs [hunk 2]
```diff
@@ -0,0 +1,562 @@
...
+    // async; decoding is what runs in parallel afterwards.
+    let mut loaded = Vec::with_capacity(partitions.len());
+    let mut reader = fileset.source.range_reader(&fileset.pgen_path).await?;
+    for partition in &partitions {
+        let mut ranges = Vec::with_capacity(partition.ranges.len());
+        for range in partition.ranges.iter().copied() {
+            let bytes = reader.read_range(range.start..range.end).await?;
+            ranges.push((range, bytes.to_vec()));
+        }
+        loaded.push(ranges);
```
[the remark this answers is not in the document]
Comment (mwiewior) [reply datafusion/bio-format-pgen/src/matrix.rs]: Fixed in 765a8d2.

The fileset is now fetched and decoded a round at a time: one byte range per partition, read concurrently, then decoded in parallel. Resident input is `partitions * max_range_bytes` rather than the whole file, and a round's reads overlap instead of queuing behind each other.

The obstacle was that the record loop carried state across the ranges it was handed — workspace, retained LD base, required-variant cursor, destination row — so `decode_partition` became a `PartitionDecoder` that keeps that state between rounds. A variant whose LD base was decoded in an earlier round still finds it.

One thing had to come with it, or the round structure would have cost ~4% at eight partitions: each partition now reads into its own reader's buffer and the decoders borrow it, so a range is not copied out at all (the previous code `to_vec()`-ed every one) and the buffers are reused round to round. With that, chromosome 22 is at parity or faster at every partition count:

| field | threads | before | after |
|---|---:|---:|---:|
| `ALT_COUNT` | 1 | 0.697 s | 0.694 s |
| `ALT_COUNT` | 8 | 0.237 s | 0.237 s |
| `DS` | 1 | 1.304 s | 1.267 s |
| `DS` | 8 | 0.380 s | 0.374 s |

`matrix_path_decodes_the_same_matrix_across_many_byte_ranges` pins the result against a range budget small enough to put nearly every record in its own range, at one, two, four and eight threads. It fails when the round loop is cut to a single round.

**Comment by mwiewior** [comment]
Review round 1 addressed — two commits.

**c6618ac — `ALT_COUNT` batch sizing** (@claude finding 1, @codex P2)
`execute_single_field` sized every fast path as `GT`, and `estimate_genotype_bytes` had no `ALT_COUNT` arm. It now sizes from the field it was handed, and `ALT_COUNT` is priced at two bytes per sample. On chr22 the scan builds **76 batches where it built 189**, at 0.60 s either way; in a multi-field projection the column is no longer priced at zero.

**765a8d2 — matrix input is streamed** (@codex P2)
The load-everything-then-decode loop held the whole compressed PGEN on top of the destination — and on a sample subset the destination is the small half — and kept object-store latency serial. It is now a round at a time: one range per partition, read concurrently, decoded in parallel, with `PartitionDecoder` carrying the per-partition state between rounds. Resident input is `partitions * max_range_bytes`.

That restructuring cost ~4% at eight partitions until the copy came out: each partition now reads into its own reader's buffer and the decoders borrow it, so a range is never copied (the previous code `to_vec()`-ed every one). With that, chr22 is at parity or faster everywhere — `DS` 1.304 → 1.267 s at one thread, 0.380 → 0.374 s at eight; `ALT_COUNT` unchanged at 0.694 s and 0.237 s.

**Nits, both taken**
- `GenotypeProjection::alt_count_only` is now `gt_only` by construction rather than by two literals that happen to agree.
- `PARALLEL_PVAR_MAX_CHUNKS` says why it is a bound on waste and deliberately not `available_parallelism`.

**Not changed**
The `decode_common_difflist_into` representation-5 invariant (@claude finding 2) and the `finish_variant` call-site pattern (finding 3) — both were assessed correct as-is in the review, and the invariant is already documented where it is load-bearing.

Verification: `cargo fmt --all -- --check`, `cargo clippy --workspace --all-targets` and the pgen suite are clean; the new multi-range matrix test fails when the round loop is cut to one round, and the ALT_COUNT sizing test fails without the new arm. Chr22 figures are means of interleaved runs, release with `-C target-cpu=native`.

@claude pls re-review
@codex pls re-review

**Comment by mwiewior** [comment]
@codex pls review
